<a href="https://colab.research.google.com/github/Pamplona-C/Notebooks/blob/master/TesteDetec%C3%A7%C3%A3oFraudes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split #Essa função é usada para dividir o conjunto de dados em duas (ou mais) partes: uma para treinamento e outra para teste
from sklearn.linear_model import LogisticRegression  #algoritmo de classificação usado para prever categorias binárias (ou múltiplas, com algumas modificações)
from sklearn.metrics import accuracy_score #mede a acurácia do modelo, que é a proporção de previsões corretas feitas pelo modelo em comparação com o total de previsões.
from sklearn.metrics import mean_squared_error #calcula a média dos quadrados das diferenças entre os valores previstos e os valores reais. Quanto menor o valor, melhor o modelo.
from sklearn.metrics import r2_score #mede o quão bem os valores previstos pelo modelo se ajustam aos dados reais.
from sklearn.metrics import confusion_matrix #mostra a relação entre as previsões feitas pelo modelo e os rótulos reais dos dados

In [3]:
#Importando o arquivo
df = pd.read_csv('creditcard.csv')

In [4]:
#verifica Data frame
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [6]:
#verifica tipos
df.dtypes

Time        int64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class     float64
dtype: object

In [7]:
#verificando se existe valores nulos
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        1
V7        1
V8        1
V9        1
V10       1
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [10]:
#observado que existe valores nulos, agora importa que ajustemos isso, trocando os valores nulos por 0
df = df.fillna(0)
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [12]:
#observado que a coluna class esta do tipo float, troquei para o tipo int
df['Class'] = df['Class'].astype(int)
df.dtypes

Time        int64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [13]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [16]:
#verifica os tipos dos dados não fraude
df_nao_fraude = df.Amount[df.Class == 0]

df_nao_fraude.describe()

count    45504.000000
mean        90.806475
std        240.320388
min          0.000000
25%          7.580000
50%         24.990000
75%         82.355000
max       7879.420000
Name: Amount, dtype: float64

Observando as transações não fraude, temos que dentro do data set temos 45504 transações não fraudulentas.
A média das transações foi 90,00 dólares,Um desvio padrão de 240, o mínimo foi 0,00, o valor máximo foi de 7.878,00 e temos os percentis onde em 25% dos casos os valores foram de 7,58 ou menos, em 50% dos casos foram valores 24,99 e 75% dos casos 82,00 ou menos

In [29]:
#verifica os tipos de dados de fraude
df_fraude = df.Amount[df.Class == 1] #relacionado a coluna amount com a coluna class onde class é igual a 1
#Vai retornar uma coluna com os valores de amount onde a coluna class é igual a 1

df_fraude.describe()

count     142.000000
mean       97.592183
std       233.185192
min         0.000000
25%         1.000000
50%         8.370000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

Observando os dados das transações fraudulentas temos que 142 transações no data set são fraudulentas, com uma média de valores de 97 dólares, um desvio padrão de 233,00 dólares, oque indica uma variabilidade alta nos valores das transações.